EIPS/eip-7748.md

```diff
@@ -0,0 +1,292 @@
+    # considering its client architecture.
+```
+
+Add or modify the following functions:
+
+```python
+# New function.
+def get_conversion_account(state: State) -> CurrentConvertingAccount:
+    # When starting the conversion, initialize with the first account
+    # in the MPT.
+    if state._conversion_curr_account is None:
+        # Initialize with the first account in the account trie.
+        first_account = trie_get_next_at_key("0x0")
+        # Accounts conversion starts with storage-slots conversion.
+        phase = StoragePhase("0x0") # Starts with the lowest storage-slot key.
+        state._conversion_curr_account = CurrentConvertingAccount(first_account, phase)
+    
+    return state._conversion_curr_account
+
+# New function.
+def conversion_move_to_next_account(state: State):
+    curr_account = state.get_conversion_account()
+    address, _, next_key = trie_get_next_at_key(state._main_trie, curr_account.phase.next_key)
+    if next_key is None:
+        # We finished the conversion
+        state._conversion_finished = True
+    else:
+        # Move to the next account
+        state._conversion_curr_account.address = address
+        state._conversion_curr_account.phase = StoragePhase("0x00")
+
+# Modified function: add new only_if_empty optional parameter.
+def set_storage(
+    state: State, addr: Address, key: Bytes, value: U256, only_if_empty: bool = True
+) -> None:
+    # <new_code>
+    if only_if_empty:
+        value = state._overlay_tree.get(get_tree_key_for_storage_slot(addr, key))
+        if value is not None:
+            return
+    # </new_code>
+    
+    state._overlay_tree.set(get_tree_key_for_storage_slot(addr, key), value)
+```
+
+As mentioned previously, the next function is called by `apply_body(...)` to perform the conversion step for a block:
+
+```python
+# Note the following function is optimized for readability, not for performance.
+def state_convert(state: State, stride: int):
+    n = 0    
+    while n < stride and not state._conversion_finished:
+        curr_account = state.get_conversion_account()
+        
+        # EIP-161 should not be converted.
+        if account_exists_and_is_empty(state, curr_account.address):
+            state.conversion_move_to_next_account()
+            continue
+        
+        # Account storage.
+        if curr_account.phase is StoragePhase:
+            # Get the storage-slot from _storage_tries which is MPT data.
+            trie = state._storage_tries.get(curr_account.address)
+            
+            if trie is not None:
+                slot_num, slot_value, next_key = trie_get_next_at_key(trie, curr_account.phase.next_key)
+                # The Overlay Tree will write in the VKT. We use the new
+                # only_if_empty parameter to avoid writing stale values.
+                set_storage(state, curr_account.address, slot_num, slot_value, only_if_empty=True)
+                n += 1
```

Comment:
Are clients able to disambiguate between a storage slot that has been (purposefully) set back to `0`, and a storage slot that has never been touched? Trying to picture an edge case where confusion would arise with regards to stale values, let's say: 
- Conversion starts at time `t`
- Slot `s` previously was assigned the value `31` in the MPT.
- Sometime after the conversion has begun, slot `s` is reset to `0`. The `0` value is then written to the verkle trie.
- At a later block, conversion reaches storage slot `s`. How can a client detect that the `0` value is not stale (and therefore should not be replaced by the `31` value, which is outdated)?

Reply:
From the EIP perspective, the `only_if_empty` is the way to signal that the write should only happen if it wasn't written by some tx. 

From a real impl perspective, it up to the client to figure this out. One option is to do a `Get` in the Verkle Tree and see if is empty or has a value. Since VKT can distinguish between empty and zero, then that's an option.  Accessing the tree can be slow so other faster options can be considered, such as some tricks using flat-dbs to write there if something was written after the conversion started, so you can assume exists in VKT and the MPT value is stale.